### DISCUSSION TITLE
SDC launch day and the Git module

### DISCUSSION THREAD

**Discussion by Dedekind561** [CodeYourFuture/curriculum#161] [Ideas]
Hello folks, 
@CodeYourFuture/syllabus-team 
We face an upcoming problem with some of the Git content in the new curriculum.

Trainees will need to submit their work with PRs against the Module-JS1 repo. To do this, they’ll need to be able to branch off main each week of the module ( same for other modules too ). 
At the moment, I think trainees will struggle with this because the current git module doesn’t tell them anything about branches in the first week. 

To fix this, we need to re-think the ordering of topics in the Git module. 
I also think there needs to be a deeper consideration of the trainee’s underlying mental model for understanding of Git from the beginning: i.e. the state of their working directory, staging area and commit history ( I appreciate this is much harder to do ). The course I’ve seen closest to this from my brief research is this: https://www.udemy.com/course/git-learning-journey/. In the course summary, she summarises exactly the objectives I think we’d have in mind for our trainees too.
I don't think we should use this course but I think we can draw heavily upon its approach in developing trainees' mental models.

**Comment by Dedekind561** [comment]
# Proposed solution
I think I can rustle up activities from existing resources we have to try and approach the introduction of Git ( Week 1 - SDC Launch day too ) in the following way. So week one would focus exclusively on the threads below:

## Week 1 ( SDC launch and Git week )
_Let’s build a mental model to understand what a repository is.
Let’s build a mental model to understand what a commit is.
Let’s build a mental model to understand why we should use version control.
Let’s figure out how to find the number of commits in a public project on Github.
Let’s figure out how to inspect the files in different commits on a public project on Github.
Let’s figure out how to get a local copy of a repository on my local machine.
Let’s figure out how to find the number of commits in a project **locally**.
Let’s figure out how to inspect the files in different commits in a project **locally**.
Let’s build a mental model to understand how we create a commit_


At the moment we leap into committing and PRs and stuff without building sufficient intuition for understanding what a commit actually is, how we can inspect timeline of commits very easily, see commits as snapshots of a repository at some point in time. In short I think they need a more tangible sense of what a commit is and what is for before they proceed.

**Reply by illicitonion** [reply]
This set of threads doesn't include anything about branches or pull requests, which it feels like they should?

Otherwise generally sounds good to me.

Some of the learning objectives I'd ~definitely want to feature:
* Fork a repository
* Clone a personal fork of a repository
* Create a local branch based on a specific upstream branch (even if not currently switched to that branch locally)
* Push a local branch to a remote fork and create a pull request containing its contents
* Solicit feedback on a pull request
* Read, understand, and respond to comments left on a pull request (both in comment form and in the form of additional content updates)
* Update a pull request with additional content
* Work on two branches in parallel and maintain two PRs independently

And some I'd consider for the into work:
* Remove accidentally committed files (i.e. via `git revert` or `git restore`, _not_ via destructive history rewrites)
* Pull in updates made upstream (i.e. `git switch main && git pull && git switch ${branch} && git pull/rebase main`)

And some things I would definitely defer until the future:
* Resolve merge conflicts

Though I mentioned a bunch of commands in this comment, I think we probably want to stick with GitHub Desktop for now over the CLI, but I could be convinced otherwise...

**Reply by Dedekind561** [reply]
Sorry I should have made it clearer that the threads I mentioned earlier are intended for week 1 only ( SDC launch week ). I think it makes sense to cover the LOs you mentioned in week 2, perhaps some overspill into week 3.

Definitely agree re merge conflicts.
I agree it makes sense not to go in with the CLI stuff at this stage. But I'm not convinced by Github Desktop either: I think the Version Control tab in VSCode makes most sense.

Reasons I'm not so keen on GH Desktop:
- I don't like the automatic staging when you click the commit button.
- You have to keep switching between 2 applications whenever you edit and save and  then inspect the diff

**Reply by illicitonion** [reply]
Aren't these all prereqs before starting any coursework-bearing elements of SDC?

I thought the HTML/CSS module coursework on github, which is why we try to teach enough GitHub to make a PR before it?

**Reply by Dedekind561** [reply]
Definitely agree re merge conflicts.
I agree it makes sense not to go in with the CLI stuff at this stage. But I'm not convinced by Github Desktop either: I think the Version Control tab in VSCode makes most sense.

Reasons I'm not so keen on GH Desktop:
- I don't like the automatic staging when you click the commit button.
- You have to keep switching between 2 applications whenever you edit and save and  then inspect the diff

**Reply by Dedekind561** [reply]
Yes they are prereqs before starting coursework so we'd need to do a bit of reordering:

## SDC launch day / Git day
Do the stuff I outlined in the first post
## Prep work for HTML week 1: what is a branch, how to branch, HTML & CSS basics
Part of their prep work before their first live session on HTML & CSS will be trying to figure out what a branch is and creating a local branch on their machine once they've cloned down the HTML & CSS coursework repo. 
## HTML week 1 Saturday session
- They can get figure out how to raise a PR together in groups ( they'd normally try and do this in on SDC launch day )
- They can do some other activities around HTML/CSS ( @SallyMcGrath I think you said you wanted trainees using Dev Tools here, right? )

**Reply by SallyMcGrath** [reply]
Yeah I've written three workshops for HTML-CSS module

**Comment by Dedekind561** [comment]
If folks are on board with this,
I think I could turn this around with 4/5 days of solid rewrite time.
In this time I would:
- sorting out Git related activities for the Saturday
- fixing the day plan
- sorting out any prep materials we need to hand out in advance

To get it done in time for WM5, I think I'd need Monday 10th to Thursday 13th of this month.

**Comment by alismith-dev** [comment]
Is it worth mentioning here that the existing [git week 1](https://syllabus.codeyourfuture.io/git/desktop/lesson#lets-do-some-homework-mentor-led-demo) has a "teacher led" demo. I'm not sure this makes as much sense in the context of flipped classroom?

I think the concept still works but just needs to be moved to a video? [I'm afraid to say that I'm on holiday next week so wouldn't be able to prepare a video before WM5]

**Reply by Dedekind561** [reply]
Yeah we'll need to update that part too, for sure.
